README.txt

Feel free to add some documentation or simply add a link to the online manual.

Visit
http://forge.typo3.org/projects/show/extension-mm_forum_news

Read the manual.

ChangeLog
10-05-24 Hauke Hain <hhpreuss@googlemail.com>

	* Read Single News PID from tt_news constant

10-04-11 Hauke Hain <hhpreuss@googlemail.com>

	* Compatibility with mm_forum_comments (using mm_forum_comments functions)

09-12-19 Martin Helmich  <m.helmich@mittwald.de>

	* First draft


lib/class.tx_mmforumnews_tcemain.php
<?php

/*                                                                     *
 *  COPYRIGHT NOTICE                                                   *
 *                                                                     *
 *  (c) 2010 Martin Helmich <m.helmich@mittwald.de>                    *
 *      All rights reserved                                            *
 *                                                                     *
 *  This script is part of the TYPO3 project. The TYPO3 project is     *
 *  free software; you can redistribute it and/or modify               *
 *  it under the terms of the GNU General Public License as published  *
 *  by the Free Software Foundation; either version 2 of the License,  *
 *  or (at your option) any later version.                             *
 *                                                                     *
 *  The GNU General Public License can be found at                     *
 *  http://www.gnu.org/copyleft/gpl.html.                              *
 *                                                                     *
 *  This script is distributed in the hope that it will be useful,     *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of     *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the      *
 *  GNU General Public License for more details.                       *
 *                                                                     *
 *  This copyright notice MUST APPEAR in all copies of the script!     *
 *                                                                     */

 require_once(t3lib_extMgm::extPath('mm_forum_comments').'lib/class.tx_mmforumcomments_div.php');


	/**
	 *
	 * This class is the main backend class of the mm_forum_news extension. It
	 * hooks into the t3lib_TCEmain class and creates a new mm_forum topic every
	 * time a tt_news record is created.
	 * If a tt_news record is edited, the post text and topic title are updated
	 * to match the news text and subject.
	 *
	 * @author    Martin Helmich <m.helmich@mittwald.de>
	 * @copyright 2009 Martin Helmich <m.helmich@mittwald.de>
	 * @version   2009-12-20
	 * @package   mm_forum_news
	 *
	 */

class tx_mmforumnews_TCEMain {



		/**
		 *
		 * Hooks into the TCEmain data-mapping process.
		 * This methods hooks into the "process_datamap" method of the TCEmain
		 * class. If a new tt_news record is updated, this function creates a
		 * mm_forum topic and stores it's UID in the new tt_news record
		 * OR
		 * the related topic is updated.
		 *
		 * @param  string        $status      The status of the edited record.
		 *                                    This may either be "new" or
		 *                                    "update".
		 * @param  string        $table       The record's table
		 * @param  string        $id          The record's UID
		 * @param  array         &$fieldArray The entire record that is being
		 *                                    edited.
		 * @param  t3lib_TCEmain $parent      The TCEmain instance that calls
		 *                                    this hook.
		 * @return void
		 */

	public function processDatamap_postProcessFieldArray($status, $table, $id, &$fieldArray, $parent) {
		if($table === 'tt_news' && $status === 'update' && intval($id) > 0) {
      $completeNewsRecord = $GLOBALS['TYPO3_DB']->sql_fetch_assoc($GLOBALS['TYPO3_DB']->exec_SELECTquery('*',$table,'uid='.$id));

      if ($fieldArray['hidden'] === null) {
        $fieldArray['hidden'] = $completeNewsRecord['hidden'];
      }

      if ($fieldArray['tx_mmforumnews_createtopic'] === null) {
        $fieldArray['tx_mmforumnews_createtopic'] = $completeNewsRecord['tx_mmforumnews_createtopic'];
      }

			if (!$fieldArray['hidden'] && $fieldArray['tx_mmforumnews_createtopic']) {
        $setup = tx_mmforumcomments_div::loadTSSetupForPage($completeNewsRecord['pid']);
    		$parameters = array('tx_ttnews->tt_news', $id, 'tx_ttnews');

    		$topicID = tx_mmforumcomments_div::getTopicID($completeNewsRecord['pid'], $parameters);

 			  if ($topicID === 0) {
          tx_mmforumcomments_div::createTopicForRecord($parameters, $setup['plugin.']['tx_mmforumcomments_pi1.'], $completeNewsRecord['pid'], $setup['plugin.']['tx_mmforum.']['storagePID'], $parent, false);
        } else {
  				//TODO: use $topicID
          /*
          $completeRecord = $GLOBALS['TYPO3_DB']->sql_fetch_assoc($GLOBALS['TYPO3_DB']->exec_SELECTquery('*',$table,'uid='.$id));
  				if($completeRecord['tx_mmforumnews_topic'])
  					$this->updateTopicForRecord($fieldArray, $completeRecord);
   				*/
  			}
      }
		}
	}

  /**
		 *
		 * Hooks into the TCEmain data-mapping process.
		 * This methods hooks into the "process_datamap" method of the TCEmain
		 * class. If a new tt_news record is created, this function creates a
		 * mm_forum topic and stores it's UID in the new tt_news record.
		 *
		 * @param  string        $status      The status of the edited record.
		 *                                    This may either be "new" or
		 *                                    "update".
		 * @param  string        $table       The record's table
		 * @param  string        $id          The record's UID
		 * @param  array         &$fieldArray The entire record that is being
		 *                                    edited.
		 * @param  t3lib_TCEmain $pObj        The TCEmain instance that calls
		 *                                    this hook.
		 * @return void
		 */

	function processDatamap_afterDatabaseOperations($status, $table, $id, &$fieldArray, &$pObj) {
		if($table === 'tt_news' && $fieldArray['tx_mmforumnews_createtopic'] && $status === 'new') {
		  $id = intval($pObj->substNEWwithIDs[$id]);

      if (intval($id) > 0 && !$fieldArray['hidden']) {
        $setup = tx_mmforumcomments_div::loadTSSetupForPage($fieldArray['pid']);
      	$parameters = array('tx_ttnews->tt_news', $id, 'tx_ttnews');

      	tx_mmforumcomments_div::createTopicForRecord($parameters, $setup['plugin.']['tx_mmforumcomments_pi1.'], $fieldArray['pid'] ,$setup['plugin.']['tx_mmforum.']['storagePID'], $parent, false);
      }
		}
	}

		/**
		 *
		 * Updates a mm_forum topic record.
		 * This function updates a mm_forum topic record, when the related
		 * tt_news record is altered. Thus, if the subject or text of the tt_news
		 * record change, the subject and the post text of the topic will be
		 * changed alongside the news record.
		 *
		 * @param array &$fieldArray    An array containing the modified fields
		 *                              in the tt_news record. The topic record
		 *                              will only be updated, if the subject or
		 *                              bodytext column of the tt_news record
		 *                              have been modified.
		 * @param array $completeRecord Since the &$fieldArray parameter contains
		 *                              only the modified columns, the complete
		 *                              and original record is submitted in this
		 *                              parameter.
		 * @return void
		 * 
		 */

	protected function updateTopicForRecord(&$fieldArray, $completeRecord) {
		global $TYPO3_DB;

		$topic = $TYPO3_DB->sql_fetch_assoc($TYPO3_DB->exec_SELECTquery('*','tx_mmforum_topics','uid='.$completeRecord['tx_mmforumnews_topic']));
		$post  = $TYPO3_DB->sql_fetch_assoc($TYPO3_DB->exec_SELECTquery('*','tx_mmforum_posts','uid='.$topic['topic_first_post_id']));

		if(isset($fieldArray['bodytext'])) {
			$TYPO3_DB->exec_UPDATEquery('tx_mmforum_posts_text',
			                            'post_id='.$post['uid'],
			                            array( 'post_text'    => $fieldArray['bodytext'],
			                                   'tstamp'       => time(),
			                                   'cache_text'   => '',
			                                   'cache_tstamp' => 0 ) );
			$TYPO3_DB->exec_UPDATEquery('tx_mmforum_posts',
			                            'uid='.$post['uid'],
				                        array( 'tstamp'     => time(),
			                                   'edit_time'  => time(),
			                                   'edit_count' => 'edit_count + 1' ),
										array( 'edit_count' ) );
		}

		if(isset($fieldArray['title']))
			$TYPO3_DB->exec_UPDATEquery('tx_mmforum_topics',
			                            'uid='.$topic['uid'],
			                            array( 'tstamp'      => time(),
			                                   'topic_title' => $fieldArray['title'] ) );

	}


		/**
		 *
		 * Instantiates and returns the mm_forum post factory. The post factory
		 * is handled as a singleton instance, i.e. it will only be instantiated
		 * ONCE and then reused.
		 *
		 * @return tx_mmforum_postfactory An instance of the tx_mmforum_postfactory
		 *                                class.
		 *
		 */

	protected function getPostFactory() {
		if(!$this->postFactory) {
			$this->postFactory = t3lib_div::getUserObj('EXT:mm_forum/pi1/class.tx_mmforum_postfactory.php:tx_mmforum_postfactory');
			$this->postFactory->init(array('storagePID' => $this->setup['plugin.']['tx_mmforum.']['storagePID']));
		} return $this->postFactory;
	}



		/**
		 *
		 * Loads the TypoScript setup for a specific page.
		 * This function loads the complete TypoScript setup for a specific
		 * page -- usually the page the tt_news record is saved on. The setup
		 * is needed in order to determine the mm_forum storage PID.
		 * The t3lib_tsparser_ext class is used for doing this.
		 *
		 * @param  int $pid The page UID for which the setup is to be loaded.
		 * @return void
		 *
		 */

	protected function loadTSSetupForPage($pid) {
		if(!$this->setup) {
			$tmpl = t3lib_div::makeInstance("t3lib_tsparser_ext");
			$tmpl->tt_track = 0;
			$tmpl->init();

			$sys_page = t3lib_div::makeInstance("t3lib_pageSelect");
			$rootLine = $sys_page->getRootLine($pid);
			$tmpl->runThroughTemplates($rootLine,0);
			$tmpl->generateConfig();
			$this->setup = $tmpl->setup;
		}
	}

}

if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/mm_forum_news/lib/class.tx_mmforumnews_tcemain.php'])
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/mm_forum_news/lib/class.tx_mmforumnews_tcemain.php']);

?>


res/ts/tx_mmforumnews_pagetsconfig.ts
mod.web_txmmforumM1 {

	defaultConfigFiles.mm_forum_news = EXT:mm_forum_news/ext_typoscript_constants.txt

	essentialConfiguration {
		mmforumnews_newsCommentCategory = 1
		mmforumnews_newsTopicAuthor     = 1
	}

	submodules {
		installation {
			categories {
				tt_news = MMFORUM_CONF_CATEGORY
				tt_news {
					icon  = EXT:mm_forum_news/res/img/mod/mmforum-conf.png
					name  = LLL:EXT:mm_forum_news/res/lang/locallang.xml:mod.category_title_short
					title = LLL:EXT:mm_forum_news/res/lang/locallang.xml:mod.category_title

					items {
						mmforumnews_newsCommentCategory = MMFORUM_CONF_ITEM
						mmforumnews_newsCommentCategory {
							type = special
							type.handler = EXT:mm_forum_news/lib/class.tx_mmforumnews_modinstall.php:tx_mmforumnews_ModInstall->getForumSelector

							label       = LLL:EXT:mm_forum_news/res/lang/locallang.xml:mod.newsCommentCategory.title
							description = LLL:EXT:mm_forum_news/res/lang/locallang.xml:mod.newsCommentCategory.desc
						}

						mmforumnews_newsTopicAuthor = MMFORUM_CONF_ITEM
						mmforumnews_newsTopicAuthor {
							type = group
							type.table = fe_users

							label       = LLL:EXT:mm_forum_news/res/lang/locallang.xml:mod.newsTopicAuthor.title
							description = LLL:EXT:mm_forum_news/res/lang/locallang.xml:mod.newsTopicAuthor.desc
						}
					}
				}
			}
		}
	}

}

lib/class.tx_mmforumnews_modinstall.php
<?php

/*                                                                     *
 *  COPYRIGHT NOTICE                                                   *
 *                                                                     *
 *  (c) 2010 Martin Helmich <m.helmich@mittwald.de>                    *
 *      All rights reserved                                            *
 *                                                                     *
 *  This script is part of the TYPO3 project. The TYPO3 project is     *
 *  free software; you can redistribute it and/or modify               *
 *  it under the terms of the GNU General Public License as published  *
 *  by the Free Software Foundation; either version 2 of the License,  *
 *  or (at your option) any later version.                             *
 *                                                                     *
 *  The GNU General Public License can be found at                     *
 *  http://www.gnu.org/copyleft/gpl.html.                              *
 *                                                                     *
 *  This script is distributed in the hope that it will be useful,     *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of     *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the      *
 *  GNU General Public License for more details.                       *
 *                                                                     *
 *  This copyright notice MUST APPEAR in all copies of the script!     *
 *                                                                     */



	/**
	 *
	 * This class provides functions for extending the mm_forum backend module.
	 *
	 * @author    Martin Helmich <m.helmich@mittwald.de>
	 * @copyright 2009 Martin Helmich <m.helmich@mittwald.de>
	 * @version   2009-12-22
	 * @package   mm_forum_news
	 *
	 */

class tx_mmforumnews_ModInstall {



		/**
		 *
		 * Generates a custom board selection field.
		 * This method generates a <SELECT> element that presents all message
		 * boards in a hierarchical order. This input field will be presented in
		 * the mm_forum configuration backend module.
		 *
		 * @author  Martin Helmich <m.helmich@mittwald.de>
		 * @version 2009-12-22
		 * @param   array  $value     The current value of this input field. The
		 *                            option with this value will be selected.
		 * @param   string $fieldname The name of the input field.
		 * @param   array  $config    A configuration array with additional
		 *                            properties for this field. In this case,
		 *                            this variable is not used.
		 * @return  string            The <SELECT> input field as HTML output.
		 *
		 */

	public function getForumSelector($value, $fieldname, $config) {
		global $TYPO3_DB;
		$content = '';
		$pid = $this->p->conf['storagePID'];

			/*
			 * Get all categories. Iterate over the categories and load the
			 * boards for each category.
			 */
		$res = $TYPO3_DB->exec_SELECTquery( 'uid, forum_name',
		                                    'tx_mmforum_forums',
			                                'parentID = 0 AND deleted=0 AND pid='.$pid, '', 'sorting ASC');
		while($arr = $TYPO3_DB->sql_fetch_assoc($res)) {
			$res2 = $TYPO3_DB->exec_SELECTquery( 'uid, forum_name',
                                                 'tx_mmforum_forums',
			                                     "parentID = {$arr[uid]} AND deleted=0 AND pid=$pid", '', 'sorting ASC' );
			if($TYPO3_DB->sql_num_rows($res2)>0) {
				$content .= '<optgroup label="'.htmlspecialchars($arr['forum_name']).'">'.chr(10);
				while($arr2 = $TYPO3_DB->sql_fetch_assoc($res2))
					$content .= '<option value="'.$arr2['uid'].'" '.($value==$arr2['uid']?'selected="selected"':'').'>'
					            . htmlspecialchars($arr2['forum_name'])
								. '</option>'.chr(10);
				$content .= '<optgroup>'.chr(10);
			}
		}

		return '<select name="tx_mmforum_install[conf][0]['.htmlspecialchars($fieldname).']">'.$content.'</select>';
	}

}

?>


ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "mm_forum_news".
#
# Auto generated 25-05-2010 01:52
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'News comments with mm_forum',
	'description' => 'Provides a commenting function for tt_news entries using the mm_forum extension.',
	'category' => 'fe',
	'shy' => 0,
	'version' => '1.0.1',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => '',
	'state' => 'beta',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearcacheonload' => 0,
	'lockType' => '',
	'author' => 'Martin Helmich, Hauke Hain',
	'author_email' => 'm.helmich@mittwald.de, hhpreuss@googlemail.com',
	'author_company' => 'Mittwald CM Service GmbH & Co. KG',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'constraints' => array(
		'depends' => array(
			'tt_news' => '3.0.1-0.0.0',
			'mm_forum' => '1.9.0-0.0.0',
			'mm_forum_comments' => '1.0.0-0.0.0',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:16:{s:9:"ChangeLog";s:4:"8bf7";s:10:"README.txt";s:4:"3981";s:12:"ext_icon.gif";s:4:"167d";s:17:"ext_localconf.php";s:4:"65b1";s:14:"ext_tables.php";s:4:"982b";s:14:"ext_tables.sql";s:4:"f0b0";s:28:"ext_typoscript_constants.txt";s:4:"5d31";s:14:"doc/manual.sxw";s:4:"3163";s:19:"doc/wizard_form.dat";s:4:"859c";s:20:"doc/wizard_form.html";s:4:"f63b";s:39:"lib/class.tx_mmforumnews_modinstall.php";s:4:"644f";s:36:"lib/class.tx_mmforumnews_tcemain.php";s:4:"5e06";s:28:"res/img/mod/mmforum-conf.png";s:4:"de58";s:22:"res/lang/locallang.xml";s:4:"bd30";s:22:"res/staticts/setup.txt";s:4:"e2e7";s:37:"res/ts/tx_mmforumnews_pagetsconfig.ts";s:4:"621b";}',
);

?>